vault/provider_test.go

```diff
@@ -618,54 +618,131 @@ func TestAccTokenName(t *testing.T) {
 							}
 						}
 					},
-					Config: testTokenNameConfig(test.UseTokenNameSchema, test.TokenNameSchema),
-					Check:  testTokenName_check(test.WantTokenName),
+					Config: testProviderConfig(test.UseTokenNameSchema, `token_name = "`+test.TokenNameSchema+`"`),
+					Check:  checkTestSecretData("display_name", test.WantTokenName),
 				},
 			},
+			CheckDestroy: testEnvCleanup("VAULT_TOKEN_NAME"),
+		})
+	}
+}
+
+func TestAccChildToken(t *testing.T) {
+
+	tests := []struct {
+		ChildTokenEnv        string
+		TestChildTokenEnv    bool
+		ChildTokenSchema     string
+		TestChildTokenSchema bool
+		ExpectChildToken     bool
+	}{
+		{
+			TestChildTokenSchema: false,
+			TestChildTokenEnv:    false,
+			ExpectChildToken:     true,
+		},
+		{
+			ChildTokenEnv:     "",
+			TestChildTokenEnv: true,
+			ExpectChildToken:  true,
+		},
+		{
+			ChildTokenEnv:     "true",
+			TestChildTokenEnv: true,
+			ExpectChildToken:  true,
+		},
+		{
+			ChildTokenEnv:     "false",
+			TestChildTokenEnv: true,
+			ExpectChildToken:  false,
+		},
+		{
+			ChildTokenSchema:     "true",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     true,
+		},
+		{
+			ChildTokenSchema:     "false",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     false,
+		},
+		{
+			ChildTokenEnv:        "true",
+			TestChildTokenEnv:    true,
+			ChildTokenSchema:     "false",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     false,
+		},
+		{
+			ChildTokenEnv:        "false",
+			TestChildTokenEnv:    true,
+			ChildTokenSchema:     "true",
+			TestChildTokenSchema: true,
+			ExpectChildToken:     true,
+		},
+	}
+
+	for _, test := range tests {
+		resource.Test(t, resource.TestCase{
+			Providers: testProviders,
+			PreCheck:  func() { testAccPreCheck(t) },
+			Steps: []resource.TestStep{
+				{
+					PreConfig: func() {
+						if test.TestChildTokenEnv {
+							err := os.Setenv("TERRAFORM_VAULT_CREATE_CHILD_TOKEN", test.ChildTokenEnv)
+							if err != nil {
+								t.Fatal(err)
+							}
+						} else {
+							err := os.Unsetenv("TERRAFORM_VAULT_CREATE_CHILD_TOKEN")
+							if err != nil {
+								t.Fatal(err)
+							}
+						}
+					},
+					Config: testProviderConfig(test.TestChildTokenSchema, `create_intermediate_child_token = `+test.ChildTokenSchema),
+					Check:  checkUsedToken(test.ExpectChildToken),
+				},
+			},
+			CheckDestroy: testEnvCleanup("TERRAFORM_VAULT_CREATE_CHILD_TOKEN"),
```

Comment:
It looks like `CheckDestroy` may be being misused here. According to the [docs](https://github.com/hashicorp/terraform-plugin-sdk/blob/76b3563a7afe40aa29d08d1661c5502b59e3811a/helper/resource/testing.go#L341-L342) it is used to ensure that the all the provisioned resources have been properly removed.

In addition `CheckDestroy` is called conditionally
[here](https://github.com/hashicorp/terraform-plugin-sdk/blob/76b3563a7afe40aa29d08d1661c5502b59e3811a/helper/resource/testing_new.go#L67-L72), so there is no guarantee that the environment will be properly cleaned. up. This could pollute the environment of other tests. I suggest adding
```
defer testEnvCleanup("TERRAFORM_VAULT_CREATE_CHILD_TOKEN")
```
 to the test function's scope.

Reply:
Thanks, that's a fair call. Moved this into the test functions as a direct env var unset.